### PR TITLE
Upgrade to Node.js 24 and finish Angular to React transition

### PR DESCRIPTION
**Pull request by guaka**
## Summary

- standardise development, CI, devcontainers and production on Node.js 24 with npm 11
- upgrade Webpack 4 to Webpack 5 and regenerate the dependency graph as a strict npm 11 lockfile
- complete the Angular-to-React transition and remove the Angular runtime, compatibility layer, routes, controllers, directives and templates
- move authentication, profiles, contacts, circles, search, offers, messages and supporting pages onto the React application shell
- preserve existing URLs through React route matching, redirects and navigation helpers
- document the supported runtime and React client architecture in OpenSpec

## Motivation

The previous application depended on an obsolete Node/npm combination, packages without Node.js 24 support, Webpack's legacy OpenSSL provider and a hybrid Angular/React client architecture.

This change establishes a current runtime and a single React client application, removing the remaining Angular bootstrap and compatibility code while retaining the existing user-facing routes and behaviour.

## Review guide

The commits are split by concern:

1. **Upgrade runtime and tooling to Node 24**
   - runtime declarations, container images, CI, dependencies, server rendering and runtime specifications

2. **Migrate the client from Angular to React**
   - React application shell, route ownership, migrated pages and removal of Angular code

3. **Restore migration behaviour and full test coverage**
   - React component, API, route, server and edge-case tests

4. **Stabilise React end-to-end coverage**
   - deterministic Playwright setup, selectors and migrated user-flow coverage

The dependency lockfile and removal of legacy Angular files account for most of the diff size.

## Acknowledgements

The React application shell and admin migration build on the foundational work in #2754 and #2755 by @awongh. That architecture and implementation helped make this complete migration possible.

## Compatibility and rollout

- Node.js 24 and npm 11 become the required development and deployment platform.
- The package lock moves to lockfile version 3.
- The browser application is React-only; Angular is no longer loaded at runtime.
- Existing public and authenticated URLs remain supported by the React router.
- Agenda remains on 1.0.3 and the direct MongoDB driver remains on 3.6.11.
- MongoDB server remains on 4.4; no database or data migration is required.
- Agenda and MongoDB driver upgrades are separated into draft follow-up #2800.
- The existing end-to-end coverage baseline is retained.

## Verification

- strict clean installation under Node.js 24 and npm 11
- full lint and formatting checks
- client: 216 suites and 1,266 tests passing
- client coverage: 100% statements, branches, functions and lines
- server coverage: 100% statements, branches, functions and lines
- Playwright: 200 tests passing
- production and development Webpack builds
- development and production Docker image builds
- GitHub Actions coverage enforcement
- OpenSpec strict validation

## Reconciliation with main

- Merged current main, preserving explicit account-deletion confirmation, profile privacy and own-account fields, username-change eligibility, and circle-only host visibility.
- Migrated the newly merged account and circle-visibility e2e tests to the React controls.
- Extracted independent messaging and signup-reminder test stability fixes into #2811. That test-only PR has merged; the runtime and React migration remain together.
- Extracted the production Docker image CI check into #2812, based directly on main; it builds without publishing the image.
- Isolated the newsletter shadowban fixture to prevent interference with profile and reference tests; extracted into #2814 for main.
- Preserved Welcome team permissions and the latest compose-area behaviour, including editor selection and immediate external resets.

Verification on `7db47fdaa`: 1,266 client tests and 1,655 server tests pass (22 existing pending), with 100% client and server coverage. Lint and both development and production image builds pass. All 200 e2e tests pass. Meeting expiry remains unchanged.

## Further independent extractions

- #2815: preserve the authenticated session when saving a profile.
- #2816: synchronise shared map coordinates and preserve user-selected zoom.
- #2817: convenient command for selected server test files.

Each targets main independently. These changes remain in this migration branch until their PRs merge into main.

- #2818: isolate e2e authentication sessions and temporary member contexts.
- #2819: additional historical spam-cleanup regression tests.

### CODE REVIEW

**Comment by awongh** [comment]
is this gonna get deployed soon? 😄

File: config/env/default.js
Comment (leon-wbr) [inline config/env/default.js:1]: Was Sentry ever used?

File: config/client/bootstrap.js
Comment (leon-wbr) [inline config/client/bootstrap.js:1]: If we are removing Bootstrap, how is design kept consistent going forward?

File: config/webpack/webpack.config.dev.js
Comment (leon-wbr) [inline config/webpack/webpack.config.dev.js:1]: I feel like Webpack files could be deduplicated a lot, but maybe the Vite transition could supersede this.

File: deploy/docker/dev-ci.Dockerfile [hunk 1]
```diff
@@ -24,15 +24,10 @@ RUN apt-get -qq update && apt-get -q install -y \
...
-# Pin npm to v7 to satisfy `engines` in package.json (`npm >=6 <8`).
-RUN npm -g i npm@latest-7
-
 WORKDIR /home/app/trustroots
 
 COPY package*.json ./
 RUN --mount=type=cache,target=/root/.npm \
-  npm ci --quiet \
-  && npm rebuild sharp --build-from-source \
-  && npm rebuild mmmagic --build-from-source
```
Comment (leon-wbr) [inline deploy/docker/dev-ci.Dockerfile:36]: Is removing the rebuild commands here safe?

File: modules/contacts/client/components/ContactAddPage.component.js
Comment (leon-wbr) [inline modules/contacts/client/components/ContactAddPage.component.js:1]: If these useEffects carry throughout the whole PR, I'm a bit skeptical. Maybe using React-Query would be a solution here. But having a mounted flag seems flimsy.

File: modules/core/client/components/TrEditor.js
Comment (leon-wbr) [inline modules/core/client/components/TrEditor.js:1]: Class-based components are not up-to-date anymore, go straight with functional.

File: modules/core/client/react-app/routes.js
Comment (leon-wbr) [inline modules/core/client/react-app/routes.js:1]: If Angular is completely gone, perhaps consider implementing a proper router makes sense at the very least. Is TanStack Router an option that we can go with for the time being that can be plugged in properly?

File: modules/core/server/services/file-upload.service.js
Comment (leon-wbr) [inline modules/core/server/services/file-upload.service.js:1]: What are these changes?

File: modules/search/client/components/SearchSidebarResults.component.js [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+  const today = new Date();
+  let age = today.getFullYear() - birth.getFullYear();
+  const monthDiff = today.getMonth() - birth.getMonth();
+
+  if (monthDiff < 0 || (monthDiff === 0 && today.getDate() < birth.getDate())) {
+    age -= 1;
+  }
+
+  return age;
+}
```
Comment (leon-wbr) [inline modules/search/client/components/SearchSidebarResults.component.js:23]: Feels like this could be simplified with some Intl thing?

File: modules/users/client/components/ProfileEditPhoto.component.js [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+export function bytesToSize(bytes) {
+  const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB'];
+
+  if (bytes === 0) {
+    return '0 Byte';
+  }
+
+  const index = parseInt(Math.floor(Math.log(bytes) / Math.log(1024)), 10);
+  return `${Math.round(bytes / 1024 ** index)} ${sizes[index]}`;
+}
```
Comment (leon-wbr) [inline modules/users/client/components/ProfileEditPhoto.component.js]: Probably should extract somewhere shared.

File: modules/users/client/utils/auth.js
Comment (leon-wbr) [inline modules/users/client/utils/auth.js:1]: I feel like Zod could help with a lot of this.

**Review by leon-wbr — APPROVED** [review]
Router is a big thing that should probably go into this PR; that will get rid of a lot of helpers and other issues. I think TanStack Router is a good way to go if it is possible to use it client-side only. But it might also be big enough to warrant a separate PR, who knows. I think it's ridiculous to hand roll routing, though.

If all of my comments are considered generally (not specific to the files where I made them), this PR will probably be ready to merge. Should be wary that the query stuff wouldn't introduce behavior differences, but I think React-Query can help to make things a better UX, as well.

File: deploy/docker/dev-ci.Dockerfile (hunk 1 above)
Comment (guaka) [reply deploy/docker/dev-ci.Dockerfile:36]: Yes. The two forced rebuilds were for sharp and mmmagic. mmmagic has been removed entirely in favour of file-type, and sharp was upgraded from 0.29.3 to 0.35.3, which installs successfully through the normal Node 24 npm ci path. I rebuilt the dev/CI image from scratch on Node 24 after removing these commands; npm ci, the full client suite and the production Webpack build all passed.

File: modules/core/client/react-app/routes.js
Comment (guaka) [reply modules/core/client/react-app/routes.js:1]: Done in e63e7e55b. The hand-written client matcher has been replaced by a TanStack Router route tree, RouterProvider and router-backed navigation. The shared server/client route policies are still the source of truth, and their paths are converted to TanStack parameter syntax so ownership, access control and rendering stay aligned. This also moves the shell to React 18 and createRoot.

File: config/webpack/webpack.config.dev.js
Comment (guaka) [reply config/webpack/webpack.config.dev.js:1]: Agreed that these configurations can be consolidated further. I would keep a Vite migration separate, though: this PR already changes the runtime, framework shell, router and Webpack 5 compatibility, while Vite would also change the build and development-serving model. The current changes are the minimum needed to keep the existing Webpack pipeline working; consolidation or Vite can follow independently.

File: config/env/default.js
Comment (guaka) [reply config/env/default.js:1]: From the repository history: Sentry was wired into both the Angular client and the Express server, but the default configuration had it disabled with an empty DSN. An external production config could have enabled it, so the repository alone cannot prove whether it was ever active in production. The removal here is intentional as part of removing the Angular integration and its legacy server bootstrap.

File: modules/core/server/services/file-upload.service.js
Comment (guaka) [reply modules/core/server/services/file-upload.service.js:1]: This replaces mmmagic, an old native addon that is incompatible with Node 24, with file-type. The service still validates the actual file contents from their magic bytes rather than trusting the uploaded MIME header. Because file-type is ESM, it is loaded with dynamic import; recognised files return the detected MIME type, unrecognised files use the existing fallback, and detection errors still reject the upload. The upload allow-list behaviour is unchanged and covered by the file-upload tests.

File: config/client/bootstrap.js
Comment (guaka) [reply config/client/bootstrap.js:1]: The file name is misleading here: config/client/bootstrap.js bootstrapped the Angular application; it was not the Bootstrap CSS framework. Bootstrap 3 LESS is still imported by config/webpack/entries/main.less, and the existing react-bootstrap components remain. So this PR preserves the current visual language. Replacing Bootstrap or introducing a new design system would be a separate change.

File: modules/core/client/components/TrEditor.js
Comment (guaka) [reply modules/core/client/components/TrEditor.js:1]: Agreed that functional components should be the default. This class is a narrow adapter around MediumEditor's imperative lifecycle and selection APIs, including the existing prop-to-editor synchronisation behaviour. Rewriting that bridge with hooks deserves an isolated change and regression coverage; the new page-level components in this PR are functional, while this adapter keeps the established editor behaviour.

File: modules/users/client/utils/auth.js
Comment (guaka) [reply modules/users/client/utils/auth.js:1]: Zod would be useful for validating untrusted API payloads, but these helpers mostly decode the existing confirmation token format, map browser validation state to messages, enforce a same-origin return path, and update authentication state. A schema would not replace most of that logic, and adding another runtime dependency solely for these helpers would not simplify this PR. The explicit checks remain covered by focused tests.

File: modules/users/client/components/ProfileEditPhoto.component.js (hunk 3 above)
Comment (guaka) [reply modules/users/client/components/ProfileEditPhoto.component.js]: I checked the merged tree and this byte-size formatter is currently used only by the profile photo editor. I would keep a single-use helper next to its consumer for now; once another upload flow needs the same presentation, extracting it into a shared file will be worthwhile.

File: modules/search/client/components/SearchSidebarResults.component.js (hunk 2 above)
Comment (guaka) [reply modules/search/client/components/SearchSidebarResults.component.js:23]: Intl does not currently calculate a person's exact calendar age. Intl.RelativeTimeFormat can format a delta after it has been calculated, but deriving the delta from milliseconds would be incorrect around birthdays and leap years. This helper intentionally calculates completed calendar years, so the explicit date comparison is still needed.

File: modules/contacts/client/components/ContactAddPage.component.js
Comment (guaka) [reply modules/contacts/client/components/ContactAddPage.component.js:1]: Agreed that this is not the ideal end state. The app still has a React Query provider, and query-backed page data should move there incrementally. In this behavioural port the local guard prevents state writes after unmount while preserving the existing request sequence; converting every ported request to React Query would be a larger data-layer change than the Angular-to-React transition itself, so I have kept that out of this PR.

File: modules/users/client/utils/auth.js
Comment (guaka) [reply modules/users/client/utils/auth.js:1]: Implemented in ecd6ffae7. Zod 4 is now used for the untrusted confirmation-token and return-path boundaries. The schemas reject malformed hex tokens and unsafe external paths while the imperative authentication-state logic stays explicit. I used Zod Mini's tree-shakeable entry and added boundary tests.

File: modules/contacts/client/components/ContactAddPage.component.js
Comment (guaka) [reply modules/contacts/client/components/ContactAddPage.component.js:1]: Implemented in ecd6ffae7. ContactAddPage now uses React Query for the member and existing-contact requests, with disabled queries for private/self cases and no mounted flag. The separate queries preserve the previous progressive loading behaviour, and the unmount cases remain covered.

File: modules/core/client/components/TrEditor.js
Comment (guaka) [reply modules/core/client/components/TrEditor.js:1]: Implemented in ecd6ffae7. The MediumEditor bridge is now a functional component using refs, effects and a layout effect for external-content synchronisation. It preserves selection, editable-input normalisation, Ctrl+Enter handling and cleanup, with focused regression tests.

File: modules/users/client/components/ProfileEditPhoto.component.js (hunk 3 above)
Comment (guaka) [reply modules/users/client/components/ProfileEditPhoto.component.js]: Implemented in ecd6ffae7. The byte formatter now lives in the shared core format utilities with dedicated tests, and the profile photo editor imports it from there.

File: config/env/default.js
Comment (guaka) [reply config/env/default.js:1]: it was actually used in the past but not in the past couple of years

**Comment by guaka** [comment]
Updated extraction overview after #2825 and #2826 merged into main; #2828 is the next extraction.

Status: 🟢 merged/already on main · 🟡 open PR · ⚪ remaining. Complexity describes migration work, not CI status.

| Group | Routes / proposed boundary | Status | Complexity |
| --- | --- | --- | --- |
| Existing React pages | Admin, support/contact, statistics and the previously migrated information pages | 🟢 Already on main | — |
| Circles | `/circles` and circle detail pages | 🟢 #2822 merged | — |
| Member entry | `/welcome`, `/navigation`, `/search/members` | 🟢 #2823 merged | — |
| Homepage and information | `/`, `/about` redirect, `/safety` | 🟢 #2824 merged | — |
| Recovery and outcome pages | `/password/forgot`, `/password/reset/success`, `/password/reset/invalid`, `/confirm-email-invalid` | 🟢 #2825 merged | Moderate |
| Contact connections | `/contact-confirm/:contactId`; `/contact-add/:userId` | 🟢 confirmation: #2826 merged; 🟡 creation: #2828 | Moderate |
| Authentication and account lifecycle | Sign-in/signup; onboarding; email-confirmation and reset-token forms; account-removal token flow | ⚪ Split entry forms from token/account lifecycle work | High |
| Profile viewing | Profile shell and about/overview, accommodation, circles, contacts and experience-history tabs | ⚪ Extract the shell with its read-only tabs | High |
| Profile editing and account settings | Editor shell with about, locations, photo and networks; then account email/username/password actions | ⚪ Two slices: profile editors, then account settings | Moderate / high |
| Writing experiences | `/profile/:username/experiences/new`; existing form in a focused recipient page | 🟡 #2828, together with contact creation | High |
| Messaging | Inbox and unread counts; then conversation threads and replies | ⚪ Two coordinated slices | High |
| Offer management | `/offer` redirect and hosting editor; meet-offer list/create/edit | ⚪ Hosting and meet offers can be separate slices | High |
| Map search | `/search`, map controls/filters, results sidebar and selected offer | ⚪ Keep the coupled map workflow together | High |
| Shared infrastructure | Runtime upgrades, Webpack/Vite work, shared service replacements and final Angular removal | ⚪ Separate tracks; #2793 informs the build-tool choice | High |

#2822, #2823, #2824, #2825 and #2826 are merged. **#2828 targets main directly**, with both contact creation and experience writing; profile viewing and experience history stay in Angular. There is no remaining stack between these extraction PRs.

#2828 preserves the existing contact/experience APIs and feature setting, named-state entry points and return links. It retains Node 16 / React 17 / Webpack 4 and refreshes the compatible IP-address helper. Three new browser tests cover guest access, edited contact submission, experience submission/duplicate detection and return to Angular history.

On #2828 head `fd49f24fd`, all 1,582 client tests pass and client/server coverage remains 100%. Lint and the Node 16 production build pass, and all 17 selected contact, recovery and experience browser checks pass. [Full GitHub CI is running](https://github.com/Trustroots/trustroots/actions/runs/34723168218). OpenSpec changes are archived.

Build-tool options and their relationship to #2769 are discussed in [#2793](https://github.com/Trustroots/trustroots/issues/2793#issuecomment-5649019001). The remaining rows above describe proposed boundaries rather than open PRs.
